packaging/release.py

```diff
@@ -0,0 +1,1410 @@
+#!/usr/bin/env python
+# PYTHON_ARGCOMPLETE_OK
+"""Manage upstream ansible-core releases."""
+
+from __future__ import annotations
+
+import argparse
+import contextlib
+import dataclasses
+import datetime
+import enum
+import functools
+import hashlib
+import http.client
+import inspect
+import json
+import math
+import os
+import pathlib
+import re
+import secrets
+import shlex
+import shutil
+import subprocess
+import sys
+import tarfile
+import tempfile
+import typing as t
+import urllib.error
+import urllib.parse
+import urllib.request
+import venv
+import webbrowser
+import zipfile
+
+import jinja2
+
+from packaging.version import Version, InvalidVersion
+
+# region CLI Framework
+
+
+C = t.TypeVar("C", bound=t.Callable[[...], None])
+
+
+def path_to_str(value: t.Any) -> str:
+    """Return the given value converted to a string suitable for use as a command line argument."""
+    return f"{value}/" if isinstance(value, pathlib.Path) and value.is_dir() else str(value)
+
+
+def run(
+    *args: t.Any,
+    env: dict[str, t.Any] | None,
+    cwd: pathlib.Path | str,
+    capture_output: bool = False,
+) -> CompletedProcess:
+    """Run the specified command."""
+    args = [arg.relative_to(cwd) if isinstance(arg, pathlib.Path) else arg for arg in args]
+
+    str_args = tuple(path_to_str(arg) for arg in args)
+    str_env = {key: path_to_str(value) for key, value in env.items()} if env is not None else None
+
+    display.show(f"--> {shlex.join(str_args)}", color=Display.CYAN)
+
+    try:
+        p = subprocess.run(str_args, check=True, text=True, env=str_env, cwd=cwd, capture_output=capture_output)
+    except subprocess.CalledProcessError as ex:
+        # improve type hinting and include stdout/stderr (if any) in the message
+        raise CalledProcessError(
+            message=str(ex),
+            args=str_args,
+            status=ex.returncode,
+            stdout=ex.stdout,
+            stderr=ex.stderr,
+        ) from None
+
+    # improve type hinting
+    return CompletedProcess(
+        args=str_args,
+        stdout=p.stdout,
+        stderr=p.stderr,
+    )
+
+
+@contextlib.contextmanager
+def suppress_when(error_as_warning: bool) -> None:
+    """Conditionally convert an ApplicationError in the provided context to a warning."""
+    if error_as_warning:
+        try:
+            yield
+        except ApplicationError as ex:
+            display.warning(ex)
+    else:
+        yield
```

Comment:
Let's dedent this a little bit:
```suggestion
    suppressed_exceptions = ApplicationError if error_as_warning else ()
    try:
        yield
    except suppressed_exceptions as ex:
        display.warning(ex)
```

Reply:
While I'm not fond of the current approach of using `yield` twice, I think I find it more readable than using a variable to hold the exception type in this case.